lightning/src/routing/scoring.rs

```diff
@@ -460,25 +453,466 @@ impl<T: Time> Readable for ChannelFailure<T> {
+	fn successful(&mut self, amount_msat: u64) {
+		let max_liquidity_msat = self.max_liquidity_msat().checked_sub(amount_msat).unwrap_or(0);
+		self.set_max_liquidity_msat(max_liquidity_msat);
+	}
+
+	/// Adjusts the lower bound of the channel liquidity balance in this direction.
+	fn set_min_liquidity_msat(&mut self, amount_msat: u64) {
+		*self.min_liquidity_offset_msat = amount_msat;
+		*self.max_liquidity_offset_msat = if amount_msat > self.max_liquidity_msat() {
+			0
+		} else {
+			self.decayed_offset_msat(*self.max_liquidity_offset_msat)
+		};
+		*self.last_updated = self.now;
+	}
+
+	/// Adjusts the upper bound of the channel liquidity balance in this direction.
+	fn set_max_liquidity_msat(&mut self, amount_msat: u64) {
+		*self.max_liquidity_offset_msat = self.capacity_msat.checked_sub(amount_msat).unwrap_or(0);
+		*self.min_liquidity_offset_msat = if amount_msat < self.min_liquidity_msat() {
+			0
+		} else {
+			self.decayed_offset_msat(*self.min_liquidity_offset_msat)
+		};
+		*self.last_updated = self.now;
+	}
+}
+
+const FRACTIONAL_BITS: u32 = 3;
+const FRACTIONAL_BITMASK: u64 = (1 << FRACTIONAL_BITS) - 1;
+const LOG2_FRACTIONAL_PART: [f64; 1 << FRACTIONAL_BITS] =
+	[0.0, 0.17, 0.32, 0.46, 0.58, 0.70, 0.81, 0.91];
+const LOG2_10: f64 = 3.32;
+
+fn log10_approx(numerator: u64, denominator: u64) -> f64 {
+	(log2_approx(numerator) - log2_approx(denominator)) / LOG2_10
+}
+
+#[inline]
+fn log2_approx(x: u64) -> f64 {
+	let leading_zeros = x.leading_zeros();
+	let integer_part = (63 - leading_zeros) as f64;
+	let fractional_part = LOG2_FRACTIONAL_PART[
+		(((x << leading_zeros) >> (63 - FRACTIONAL_BITS)) & FRACTIONAL_BITMASK) as usize
+	];
+	integer_part + fractional_part
+}
+
+fn log10_approx_without_fractional(numerator: u64, denominator: u64) -> f64 {
+	(log2_approx_without_fractional(numerator) - log2_approx_without_fractional(denominator)) / LOG2_10
+}
+
+#[inline]
+fn log2_approx_without_fractional(x: u64) -> f64 {
+	(63 - x.leading_zeros()) as f64
+}
+
+impl<G: Deref<Target = NetworkGraph>, T: Time> Score for ProbabilisticScorerUsingTime<G, T> {
+	fn channel_penalty_msat(
+		&self, short_channel_id: u64, amount_msat: u64, capacity_msat: u64, source: &NodeId,
+		target: &NodeId
+	) -> u64 {
+		if *source == self.node_id || *target == self.node_id {
+			return 0;
+		}
+
+		let liquidity_penalty_multiplier_msat = self.params.liquidity_penalty_multiplier_msat;
+		let liquidity_offset_half_life = self.params.liquidity_offset_half_life;
+		let success_probability = self.channel_liquidities
+			.get(&short_channel_id)
+			.unwrap_or(&ChannelLiquidity::new())
+			.as_directed(source, target, capacity_msat, liquidity_offset_half_life)
+			.success_probability(amount_msat);
+		match success_probability {
+			Probability::Zero => u64::max_value(),
+			Probability::One => 0,
+			Probability::Ratio { numerator, denominator } => {
+				let log_success_probability = log10_approx(numerator, denominator);
```

Comment:
While the log is merely introduced to switch from the multiplicative nature of channel probabilities for an entire path to adding the scores per hop and making this metric usable for shortest path algos and min cost flows the base of the log does not really matter. for the actual computation / ranking of the scored paths. Of corse it might when combinging with other features in a weighted arithmetic mean...

However from an information theoretic point of view the base 2 log is usually used as that would measure the entropy and information gain in bits. it would also allow us to have very fast approximations as we just needed the first significant bit of a number.

Reply:
> While the log is merely introduced to switch from the multiplicative nature of channel probabilities for an entire path to adding the scores per hop and making this metric usable for shortest path algos and min cost flows the base of the log does not really matter. for the actual computation / ranking of the scored paths. Of corse it might when combinging with other features in a weighted arithmetic mean...
> 

Note earlier comment about how we combine penalty and fees to get the cost.

> However from an information theoretic point of view the base 2 log is usually used as that would measure the entropy and information gain in bits. it would also allow us to have very fast approximations as we just needed the first significant bit of a number.

IIRC, when using base 2, I would have a zero penalty for larger success probabilities when base 10 would give me something small but non-zero. Though this may also have had to due with the value I was multiplying by.

Additionally, when using `log2(numerator) - log2(denominator)`, it's easy to get a probability of zero if the amount is  close to the min liquidity. That is, subtracting them from max liquidity to get the numerator and denominator, respectively, may give two values with identical most significant bits (i.e., zero when subtracted).

Reply:
> > While the log is merely introduced to switch from the multiplicative nature of channel probabilities for an entire path to adding the scores per hop and making this metric usable for shortest path algos and min cost flows the base of the log does not really matter. for the actual computation / ranking of the scored paths. Of corse it might when combinging with other features in a weighted arithmetic mean...
> 
> Note earlier comment about how we combine penalty and fees to get the cost.

note my reply: https://github.com/lightningdevkit/rust-lightning/pull/1227#discussion_r789270476 I advice against the way how you currently do that.

> IIRC, when using base 2, I would have a zero penalty for larger success probabilities when base 10 would give me something small but non-zero. Though this may also have had to due with the value I was multiplying by.

nit: which - when rounding is taking place - might actually be ok. small payments kind of don't mind which channel to take. They have a really high probability and the log could be close to 0 or 0 (though technically only for probability of 1 we want zero as a result.

> Additionally, when using `log2(numerator) - log2(denominator)`, it's easy to get a probability of zero if the amount is close to the min liquidity. That is, subtracting them from max liquidity to get the numerator and denominator, respectively, may give two values with identical most significant bits (i.e., zero when subtracted).

Shouldn't that only be true if you take integers of the base2 log or like I suggested estimate this with the position of the first non zero bit of the `numerator` and `denominator`?